python/tests/expression/test_literals.py

```diff
@@ -0,0 +1,665 @@
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import math
+import uuid
+from datetime import datetime
+from decimal import Decimal
+
+import dateutil.parser
+import pytest
+from fastavro.write import LOGICAL_WRITERS
+from pytest import raises
+
+from iceberg.expression.literals import (
+    EPOCH,
+    AboveMax,
+    BaseLiteral,
+    BelowMin,
+    ComparableLiteral,
+    DateLiteral,
+    Literal,
+    StringLiteral,
+    TimestampLiteral,
+    literal,
+)
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+# Base
+
+
+def test_literal_from_none_error():
+    with pytest.raises(TypeError) as e:
+        literal(None)
+    assert "Unimplemented Type Literal for value" in str(e.value)
+
+
+def test_string_literal_with_none_value_error():
+    with pytest.raises(TypeError) as e:
+        StringLiteral(None)
+    assert "Cannot set value of BaseLiteral to None" in str(e.value)
+
+
+@pytest.mark.parametrize(
+    "clazz",
+    [
+        Literal(),
+        BaseLiteral("baseLiteral", 123),
+        ComparableLiteral("comparableLiteral", 456),
+    ],
+)
+def test_base_class_not_implemented_error(clazz):
+    with pytest.raises(NotImplementedError):
+        clazz.to(IntegerType())
+
+
+def test_base_literal():
+    b = BaseLiteral("baseLiteral", 3.14)
+    assert repr(b) == "baseLiteral"
+    assert str(b) == "3.14"
+    assert b.value == 3.14
+
+
+# Numeric
+
+
+def test_numeric_literal_comparison():
+    small_lit = literal(10).to(IntegerType())
+    big_lit = literal(1000).to(IntegerType())
+    assert small_lit != big_lit
+    assert small_lit == literal(10)
+    assert small_lit < big_lit
+    assert small_lit <= big_lit
+    assert big_lit > small_lit
+    assert big_lit >= small_lit
+
+
+def test_integer_to_long_conversion():
+    lit = literal(34)
+    long_lit = lit.to(LongType())
+
+    assert lit.value == long_lit.value
+
+
+def test_integer_to_float_conversion():
+    lit = literal(34)
+    float_lit = lit.to(FloatType())
+
+    assert math.isclose(lit.value, float_lit.value)
+
+
+def test_integer_to_double_conversion():
+    lit = literal(34)
+    dbl_lit = lit.to(DoubleType())
+
+    assert math.isclose(lit.value, dbl_lit.value)
+
+
+@pytest.mark.parametrize(
+    "decimalType, decimalValue", [(DecimalType(9, 0), "34"), (DecimalType(9, 2), "34.00"), (DecimalType(9, 4), "34.0000")]
+)
+def test_integer_to_decimal_conversion(decimalType, decimalValue):
+    lit = literal(34)
+
+    assert lit.to(decimalType).value.as_tuple() == Decimal(decimalValue).as_tuple()
+
+
+def test_integer_to_date_conversion():
+    lit = literal(20220303)
```

Comment:
This is a bit misleading because the integer here looks like a date. But this should actually represent 20 million days from the unix epoch. What this should test is that the integer and literal value are the same object: `lit.value == date_lit.value` and the result is a `DateLiteral`.